torch/nn/parallel/distributed.py

```diff
@@ -222,21 +222,27 @@ def __init__(self, module, device_ids=None,
         )
 
         self.is_multi_device_module = len({p.device for p in module.parameters()}) > 1
-        self.is_cuda = all([p.device.type == 'cuda' for p in module.parameters()])
-
-        if not self.is_cuda or self.is_multi_device_module:
+        device_type = {p.device.type for p in module.parameters()}
+        assert len(device_type) == 1, (
+            "DistributedDataParallel with multi-device module only works "
+            "with GPU devices, but module parameters locate in {}."
```

Comment:
This error message does not seem to be accurate? This seems to assert that "DistributedDataParallel's input module must locate on the same type of device, but the provided module parameters locate in {}"? Or maybe I misunderstood it.

Reply:
Yes, your understanding is correct.
The restrictions are:
1. Multiple-device module must be on the same device type
2. Multiple-device module must be distributed as **"single-process-single-device"** (Here I mean the user cannot replicate the input multiple-device module in single process with DDP)

The check here is to guarantee the first limitation.